rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,321 @@
+#[derive(Debug)]
+pub struct Diagnostic<'pr> {
+    diagnostic: NonNull<yp_diagnostic_t>,
+    marker: PhantomData<&'pr yp_diagnostic_t>
+}
+
+impl<'pr> Diagnostic<'pr> {
+    /// Returns the message associated with the diagnostic.
+    /// 
+    /// # Panics
+    ///
+    /// Panics if the message is not able to be converted into a `CStr`.
+    ///
+    #[must_use]
+    pub fn message(&self) -> &str {
+        unsafe {
+            let message: *mut c_char = self.diagnostic.as_ref().message as *mut c_char;
+            CStr::from_ptr(message).to_str().expect("YARP allows only UTF-8 for diagnostics.")
+        }
+    }
+}
+
+/// A comment that was found during parsing.
+#[derive(Debug)]
+pub struct Comment<'pr> {
+    comment: NonNull<yp_comment_t>,
+    marker: PhantomData<&'pr yp_comment_t>
+}
+
+impl<'pr> Comment<'pr> {
+    /// Returns the text of the comment.
+    ///
+    /// # Panics
+    /// 
+    /// Panics if the comment is not able to be read.
+    /// 
+    #[must_use]
+    pub fn text(&self) -> &[u8] {
+        unsafe {
+            let start = self.comment.as_ref().start.cast::<u8>();
+            let end = self.comment.as_ref().end.cast::<u8>();
+
+            let len = usize::try_from(end.offset_from(start)).expect("end should point to memory after start");
+            std::slice::from_raw_parts(start, len)
+        }
+    }
+}
+
+/// A struct created by the `errors` or `warnings` methods on `ParseResult`. It
+/// can be used to iterate over the diagnostics in the parse result.
+pub struct Diagnostics<'pr> {
+    diagnostic: *mut yp_diagnostic_t,
+    marker: PhantomData<&'pr yp_diagnostic_t>
+}
+
+impl<'pr> Iterator for Diagnostics<'pr> {
+    type Item = Diagnostic<'pr>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.diagnostic.is_null() {
+            None
+        } else {
+            let current = Diagnostic {
+                diagnostic: unsafe { NonNull::new_unchecked(self.diagnostic) },
+                marker: PhantomData
+            };
+
+            self.diagnostic = unsafe { (*(self.diagnostic)).node.next.cast::<yp_diagnostic_t>() };
+            Some(current)
+        }
+    }
+}
+
+/// A struct created by the `comments` method on `ParseResult`. It can be used
+/// to iterate over the comments in the parse result.
+pub struct Comments<'pr> {
+    comment: *mut yp_comment_t,
+    marker: PhantomData<&'pr yp_comment_t>
+}
+
+impl<'pr> Iterator for Comments<'pr> {
+    type Item = Comment<'pr>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match NonNull::new(self.comment) {
+            None => None,
+            Some(comment) => {
+                let current = Comment { comment, marker: PhantomData };
+                self.comment = unsafe { comment.as_ref().node.next.cast::<yp_comment_t>() };
+                Some(current)
+            }
+        }
+    }
+}
+
+/// The result of parsing a source string.
+pub struct ParseResult<'pr> {
```

Comment:
this should derive Debug. Rust API guidelines suggest all public structs derive Debug.